traffic_ops/traffic_ops_golang/deliveryservice/acme.go

```diff
@@ -0,0 +1,363 @@
+	}
+	storedAcmeInfo, err := getStoredAcmeAccountInfo(userTx, acmeAccount.UserEmail, acmeAccount.AcmeProvider)
+	if err != nil {
+		log.Errorf("Error finding stored ACME information: %s", err.Error())
+		return nil, err
+	}
+
+	myUser := MyUser{}
+	foundPreviousAccount := false
+	userPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		log.Errorf("Error generating private key: %s", err.Error())
+		return nil, err
+	}
+
+	if storedAcmeInfo == nil || acmeAccount.UserEmail == "" {
+		myUser = MyUser{
+			key:   userPrivateKey,
+			Email: acmeAccount.UserEmail,
+		}
+	} else {
+		foundPreviousAccount = true
+		myUser = MyUser{
+			key:   &storedAcmeInfo.PrivateKey,
+			Email: storedAcmeInfo.Email,
+			Registration: &registration.Resource{
+				URI: storedAcmeInfo.URI,
+			},
+		}
+	}
+
+	config := lego.NewConfig(&myUser)
+	config.CADirURL = acmeAccount.AcmeUrl
+	config.Certificate.KeyType = certcrypto.RSA2048
+
+	client, err := lego.NewClient(config)
+	if err != nil {
+		log.Errorf("Error creating acme client: %s", err.Error())
+		return nil, err
+	}
+
+	if acmeAccount.AcmeProvider == tc.LetsEncryptAuthType {
+		client.Challenge.Remove(challenge.HTTP01)
+		client.Challenge.Remove(challenge.TLSALPN01)
+		trafficRouterDns := NewDNSProviderTrafficRouter()
+		trafficRouterDns.db = db
+		if err != nil {
+			log.Errorf("Error creating Traffic Router DNS provider: %s", err.Error())
+			return nil, err
+		}
+		client.Challenge.SetDNS01Provider(trafficRouterDns)
+	}
+
+	if foundPreviousAccount {
+		log.Debugf("Found existing account with %s", acmeAccount.AcmeProvider)
+		reg, err := client.Registration.QueryRegistration()
+		if err != nil {
+			log.Errorf("Error querying %s for existing account: %s", acmeAccount.AcmeProvider, err.Error())
+			return nil, err
+		}
+		myUser.Registration = reg
+		if reg.Body.Status != "valid" {
+			log.Debugf("Account found with %s is not valid.", acmeAccount.AcmeProvider)
+			foundPreviousAccount = false
+		}
+	}
+	if !foundPreviousAccount {
+		if acmeAccount.Kid != "" && acmeAccount.HmacEncoded != "" {
+			reg, err := client.Registration.RegisterWithExternalAccountBinding(registration.RegisterEABOptions{
+				TermsOfServiceAgreed: true,
+				Kid:                  acmeAccount.Kid,
+				HmacEncoded:          acmeAccount.HmacEncoded,
+			})
+			if err != nil {
+				log.Errorf("Error registering acme client with external account binding: %s", err.Error())
+				return nil, err
+			}
+			myUser.Registration = reg
+			log.Debugf("Creating a new account with %s", acmeAccount.AcmeProvider)
+		} else {
+			reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
+			if err != nil {
+				log.Errorf("Error registering acme client: %s", err.Error())
+				return nil, err
+			}
+			myUser.Registration = reg
+			log.Debugf("Creating a new account with %s", acmeAccount.AcmeProvider)
+		}
+
+		// save account info
+		userKeyDer := x509.MarshalPKCS1PrivateKey(userPrivateKey)
```

Comment:
Do you think we should separate the key generation/ renewal part into its own function?

Reply:
yea thats a good idea. moved it and called it from the LE code too